tests/snappi_tests/multidut/pfc/files/m2o_oversubscribe_lossless_lossy_helper.py

```diff
@@ -0,0 +1,334 @@
+# uncompyle6 version 3.9.0
+# Python bytecode version base 2.7 (62211)
+# Decompiled from: Python 3.10.4 (tags/v3.10.4:9d38120, Mar 23 2022, 23:13:41) [MSC v.1929 64 bit (AMD64)]
+# Embedded file name: /var/johnar/sonic-mgmt/tests/snappi/multi_dut_rdma/files/rdma_helper.py
+# Compiled at: 2023-02-10 09:15:26
+from math import ceil                                                                   # noqa: F401
+import logging                                                                          # noqa: F401
+from tests.common.helpers.assertions import pytest_assert, pytest_require               # noqa: F401
+from tests.common.fixtures.conn_graph_facts import conn_graph_facts, fanout_graph_facts  # noqa: F401
+from tests.common.snappi_tests.snappi_helpers import get_dut_port_id                     # noqa: F401
+from tests.common.snappi_tests.common_helpers import pfc_class_enable_vector, \
+    stop_pfcwd, disable_packet_aging, sec_to_nanosec                                     # noqa: F401
+from tests.common.snappi_tests.port import select_ports                                 # noqa: F401
+from tests.common.snappi_tests.snappi_test_params import SnappiTestParams
+from tests.common.snappi_tests.traffic_generation import run_traffic, \
+     setup_base_traffic_config, verify_m2o_oversubscribtion_results                      # noqa: F401
+
+logger = logging.getLogger(__name__)
+
+PAUSE_FLOW_NAME = 'Pause Storm'
+TEST_FLOW_NAME = 'Test Flow'
+TEST_FLOW_AGGR_RATE_PERCENT = [40, 20]
+BG_FLOW_NAME = 'Background Flow'
+BG_FLOW_AGGR_RATE_PERCENT = [20, 40]
+DATA_PKT_SIZE = 1024
+DATA_FLOW_DURATION_SEC = 20
+DATA_FLOW_DELAY_SEC = 10
+SNAPPI_POLL_DELAY_SEC = 2
+TOLERANCE_THRESHOLD = 0.05
+
+
```

Comment:
please add comments on how to test is done and expected results.

Reply:
The test data traffic will include two lossless traffic streams, with the SONiC default lossless priorities of 3 and 4, and first stream having a 40% bandwidth from port 1 and second stream having 20% of bandwidth from port 2.
The background traffic will consist of two lossy traffic streams, each with randomly chosen priorities (0..2, 5..7), and one stream to be at 20% from port 1while the other is at 40% bandwidth from port 2. Start the test data traffic from IXIA Tx ports 1; the combined line rate will be 60% of the egress port capacity.
Start the background traffic from IXIA Tx ports 2; the additional increase in bandwidth will be 60%, for a total oversubscription of 120%. After a few seconds, measure the received bandwidth per flow on the IXIA Rx port 3: test data traffic and background traffic with 20% of bandwidth should not lose any packet. Internally the router should be dropping lossy frames from the streams causing the over-subscription (Port 2). There should be PFC pause messages being received at the IXIA transmit port 1 stream 2 causing the over-subscription.

Reply:
The background traffic will include two lossy traffic streams, each with randomly chosen priorities (0..2, 5..7), and each having a 25% bandwidth. The test data traffic will consist of two lossless traffic streams, with the SONiC default lossless priorities of 3 and 4, and each having a 25% bandwidth. Start the background and test traffic from IXIA Tx ports 1 & 2; the combined line rate will be 100% of the egress port capacity (25% line rate bandwidth per stream)After a few seconds, start the PFC pause storm to fully block one or all test priorities at the switch. Ensure that: a. The IXIA Rx port is receiving all lossy streams and any of the lossless streams not marked for PFC pause due to randomization. b. The IXIA Rx port does not receive any of the lossless streams marked to be fully blocked by PFC